### ISSUE TITLE
Composite Decorators Defining Metadata on Class when Targeting Methods

### ISSUE THREAD

**Issue by mkaufmaner** [nestjs/nest#4039]
## Current behavior
Currently, when using the `applyDecorators` https://docs.nestjs.com/custom-decorators#decorator-composition with static class methods the decorators are applied to the class instead of the method.

## Input Code
```ts
/* eslint-disable max-classes-per-file */
import { GUARDS_METADATA } from '@nestjs/common/constants';
import { applyDecorators, UseGuards, CanActivate, Controller } from '@nestjs/common';

class Guard implements CanActivate {
	canActivate() {
		return true;
	}
}

const GuardCompositeDecorator = () => {
	return applyDecorators(UseGuards(Guard));
};

describe('UseGuards', () => {
	@Controller()
	@UseGuards(Guard)
	class Test {}

	@Controller()
	class TestWithMethod {
		@UseGuards(Guard)
		public test() {
			return true;
		}
	}

	@Controller()
	class TestWithStaticMethod {
		@UseGuards(Guard)
		public static test() {
			return true;
		}
	}

	it('should be using the guard defined on the class', () => {
		const classMetadata = Reflect.getMetadata(GUARDS_METADATA, Test);
		expect(classMetadata).toEqual([Guard]);
	});

	it('should be using the guard defined on the prototype method', () => {
		const instance = new TestWithMethod();

		const classMetadata = Reflect.getMetadata(GUARDS_METADATA, TestWithMethod);
		const methodMetadata = Reflect.getMetadata(GUARDS_METADATA, instance.test);
		const instanceMetadata = Reflect.getMetadata(GUARDS_METADATA, instance);

		expect(classMetadata).toBeUndefined();
		expect(methodMetadata).toEqual([Guard]);
		expect(instanceMetadata).toBeUndefined();
	});

	it('should be using the guard defined on the static method', () => {
		const classMetadata = Reflect.getMetadata(GUARDS_METADATA, TestWithStaticMethod);
		const methodMetadata = Reflect.getMetadata(GUARDS_METADATA, TestWithStaticMethod.test);

		expect(classMetadata).toBeUndefined();
		expect(methodMetadata).toEqual([Guard]);
	});
});

describe('@GuardCompositeDecorator', () => {
	@Controller()
	@GuardCompositeDecorator()
	class Test {}

	@Controller()
	class TestWithMethod {
		@GuardCompositeDecorator()
		public test() {
			return true;
		}
	}

	@Controller()
	class TestWithStaticMethod {
		@GuardCompositeDecorator()
		public static test() {
			return true;
		}
	}

	it('should be using the guard defined on the class', () => {
		const classMetadata = Reflect.getMetadata(GUARDS_METADATA, Test);
		expect(classMetadata).toEqual([Guard]);
	});

	it('should be using the guard defined on the prototype method', () => {
		const instance = new TestWithMethod();

		const classMetadata = Reflect.getMetadata(GUARDS_METADATA, TestWithMethod);
		const methodMetadata = Reflect.getMetadata(GUARDS_METADATA, instance.test);
		const instanceMetadata = Reflect.getMetadata(GUARDS_METADATA, instance);

		expect(classMetadata).toBeUndefined();
		expect(methodMetadata).toEqual([Guard]);
		expect(instanceMetadata).toBeUndefined();
	});

	it('should be using the guard defined on the static method', () => {
		const classMetadata = Reflect.getMetadata(GUARDS_METADATA, TestWithStaticMethod);
		const methodMetadata = Reflect.getMetadata(GUARDS_METADATA, TestWithStaticMethod.test);

		expect(classMetadata).toBeUndefined();
		expect(methodMetadata).toEqual([Guard]);
	});
});
```

Output:
```bash
jest ./bad.decorator.spec.ts
 FAIL  bad.decorator.spec.ts
  UseGuards
    ✓ should be using the guard defined on the class (3ms)
    ✓ should be using the guard defined on the prototype method (1ms)
    ✓ should be using the guard defined on the static method
  @GuardCompositeDecorator
    ✓ should be using the guard defined on the class
    ✓ should be using the guard defined on the prototype method (1ms)
    ✕ should be using the guard defined on the static method (2ms)

  ● @GuardCompositeDecorator › should be using the guard defined on the static method

    expect(received).toBeUndefined()

    Received: [[Function Guard]]

      102 | 		const methodMetadata = Reflect.getMetadata(GUARDS_METADATA, TestWithStaticMethod.test);
      103 | 
    > 104 | 		expect(classMetadata).toBeUndefined();
          | 		                      ^
      105 | 		expect(methodMetadata).toEqual([Guard]);
      106 | 	});
      107 | });

      at Object.<anonymous> (bad.decorator.spec.ts:104:25)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 5 passed, 6 total
Snapshots:   0 total
Time:        1.037s
```

## Expected behavior
I would expect the static method to be defined with metadata.

## Possible Solution

Not 100% sure, but I believe this has to do with the `apply-decorators.ts` logic determines the target of the decorator `if (target instanceof Function) {`.
```ts
/**
 * Function that returns a new decorator that applies all decorators provided by param
 *
 * Useful to build new decorators (or a decorator factory) encapsulating multiple decorators related with the same feature
 *
 * @param decorators one or more decorators (e.g., `ApplyGuard(...)`)
 *
 * @publicApi
 */
export function applyDecorators(
  ...decorators: Array<ClassDecorator | MethodDecorator | PropertyDecorator>
) {
  return <TFunction extends Function, Y>(
    target: TFunction | object,
    propertyKey?: string | symbol,
    descriptor?: TypedPropertyDescriptor<Y>,
  ) => {
    for (const decorator of decorators) {
      if (target instanceof Function) {
        (decorator as ClassDecorator)(target);
        continue;
      }
      (decorator as MethodDecorator | PropertyDecorator)(
        target,
        propertyKey,
        descriptor,
      );
    }
  };
}

```

## Environment

<pre><code>
Nest version: ^6.11.5

 
For Tooling issues:
- Node version: 12.13.0
- Platform:  N/A

Others:

</code></pre>

**Comment by mkaufmaner** [comment]
Related to https://github.com/nestjs/nest/issues/3104

**Comment by kamilmysliwiec** [comment]
Thanks for reporting. Can you test with 6.11.7? Also, why do you need to apply decorators on static methods (and which NestJS decorators you're willing to use)

**Comment by mkaufmaner** [comment]
@kamilmysliwiec I am not sure exactly why anyone would apply this against a static method in practice, I was more or less following the existing NestJS tests for other decorators when I stumbled upon this.

Check out https://github.com/nestjs/nest/blob/master/packages/common/test/decorators/use-guards.decorator.spec.ts#L16

This leads me to believe that there should NOT be any logic within `apply-decorators.ts` to determine the whether the decorator is being applied to a class or method.  Instead, I recommend a mixin to be used similar to how the `@SetMetadata` decorator works.

```ts
/**
 * Function that returns a new decorator that applies all decorators provided by param
 *
 * Useful to build new decorators (or a decorator factory) encapsulating multiple decorators related with the same feature
 *
 * @param decorators one or more decorators (e.g., `ApplyGuard(...)`)
 *
 * @publicApi
 */
export function applyDecorators(
  ...decorators: Array<ClassDecorator & MethodDecorator & PropertyDecorator>
) {
  return <TFunction extends Function, Y>(
    target: TFunction | object,
    propertyKey?: string | symbol,
    descriptor?: TypedPropertyDescriptor<Y>,
  ) => {
    for (const decorator of decorators) {
      (decorator)(
        target,
        propertyKey,
        descriptor
      );
    }
  };
}
```

**Comment by mkaufmaner** [comment]
Looks like you fixed it in 6.7.11! Thank you, and awesome! https://github.com/nestjs/nest/commit/1cffa700a94972ece43e339b89eb1de5a20e2202

Would still like to know your thoughts on making it a mixin and leaving the logic to determine the type of decorator up to the decorators that were wrapped.
